src/v/archival/ntp_archiver_service.cc

```diff
@@ -232,6 +232,8 @@ ss::future<> ntp_archiver::upload_loop() {
             break;
         }
 
+        update_probe();
```

Comment:
This location means that we'll only update the stats on the leader + leave behind some incorrect stats when we lose leadership: perhaps we should update stats when applying updates to archival_metadata_stm instead?

Alternatively (albeit a bit more invasively) we could register+deregister stats when we gain/lose leadership.  I would actually prefer that, but haven't looked into the complexity.

Reply:
I'm not sure I see how we can lose leadership and still publish data in this block. `upload_loop_can_continue` checks that we're the leader and there's not scheduling points between it and `update_probe`.
https://github.com/redpanda-data/redpanda/blob/4c8d532c521e2bc5b894482f9d7d937cdfa91e1b/src/v/archival/ntp_archiver_service.cc#L231-L235

Reply:
I tied the probe lifetime to the upload loop (I don't really want to thread the leaders table through to each ntp archiver object.). It can still get a bit out of sync if the leadership change happens while the upload loop is backing off, but it will correct itself after a maximum of 10 seconds (the default max back-off for the upload loop).